includes/reader-revenue/woocommerce/class-woocommerce-cli.php

```diff
@@ -162,6 +162,133 @@ function( $subscription ) {
+		'
+		);
+
+		$query_args    = [
+			'subscriptions_per_page' => $batch_size,
+			'subscription_status'    => [ 'active', 'pending' ],
+			'offset'                 => 0,
+		];
+		$processed     = 0;
+		$subscriptions = \wcs_get_subscriptions( $query_args );
+		$total_revenue = 0;
+		$results       = [];
+
+		while ( ! empty( $subscriptions ) ) {
+			foreach ( $subscriptions as $subscription_id => $subscription ) {
+				array_shift( $subscriptions );
+				$subscription_start = $subscription->get_date( 'start_date' );
+
+				// If the subscription start date is before the $args start date, we're done.
+				if ( strtotime( $subscription_start ) < $start_date ) {
+					$subscriptions = [];
+					break;
+				}
+
+				$next_payment_date = $subscription->get_date( 'next_payment' );
+				$is_in_past        = ! strtotime( $next_payment_date ) || strtotime( $next_payment_date ) < $now;
+
+				// Subscription has a valid next payment date and it's in the future, so skip.
+				if ( $next_payment_date && ! $is_in_past ) {
+					continue;
+				}
+
+				$result   = [
+					'ID'                => $subscription->get_id(),
+					'status'            => $subscription->get_status(),
+					'start_date'        => $subscription_start,
+					'next_payment_date' => $next_payment_date,
+					'billing_period'    => $subscription->get_billing_period(),
+					'missed_periods'    => 0,
+					'missed_total'      => 0,
+				];
+				$min_date = strtotime( $subscription_start );
+				while ( $min_date <= $now ) {
+					$result['missed_periods']++;
+					$min_date = strtotime( '+1 ' . $result['billing_period'], $min_date );
+				}
+
+				if ( $result['missed_periods'] ) {
+					$result['missed_total'] += $subscription->get_total() * $result['missed_periods'];
+					$total_revenue += $result['missed_total'];
+				}
+
+				if ( ! $dry_run ) {
+					$subscription->update_dates(
+						[
+							'next_payment' => $subscription->calculate_date( 'next_payment' ),
+						]
+					);
+					$subscription->save();
+					$result['next_payment_date'] = $subscription->get_date( 'next_payment' );
+				}
+
+				$results[] = $result;
+				$processed++;
+
+				// Get the next batch.
+				if ( empty( $subscriptions ) ) {
+					$query_args['offset'] += $batch_size;
+					$subscriptions        = \wcs_get_subscriptions( $query_args );
+				}
+			}
+		}
+
+		if ( empty( $results ) ) {
+			\WP_CLI::log( 'No subscriptions with missing next_payment dates found in the given time period.' );
+		} else {
+			\WP_CLI\Utils\format_items(
+				'table',
+				$results,
+				[
+					'ID',
+					'status',
+					'start_date',
+					'next_payment_date',
+					'billing_period',
+					'missed_periods',
+					'missed_total',
+				]
+			);
+			\WP_CLI::success(
+				sprintf(
+					'Finished processing %d subscriptions. %s',
+					$processed,
+					$total_revenue ? 'Total missed revenue: ' . \wp_strip_all_tags( html_entity_decode( \wc_price( $total_revenue ) ) ) : ''
```

Comment:
This is not exactly true: if a sub is starting _now_ (had one payment), and does not have next payment date, no revenue was missed (it will be _after_ the next billing period).

Reply:
Good point! 6e5c96f4 changes this so that we don't start counting missed periods until _after_ the first one. I think this should be more accurate, right?

Reply:
Makes sense!